### PR TITLE
Allow server_url without port and https prefix.

### PR DESCRIPTION
**Pull request by masteryyh**
**Related issue**
[#2737](https://github.com/harvester/harvester/issues/2737)

**Test steps**
1. Build an ISO with this branch;
2. Use the artifacts built with this branch to construct a PXE server, configure a `join` configuration and use a server URL without port, like `https://172.16.0.240`;
3. Use the PXE server to boot up a node and start automatic installation;
4. Complete the installation and see if node successfully joined cluster;
5. Get into console, run `yq '.serverurl' < /oem/harvester.config` to see if server URL is correctly configured.

Signed-off-by: Yanhong Yang <yanhong.yang@suse.com>

### CODE REVIEW

File: pkg/console/util.go [hunk 1]
```diff
@@ -277,12 +278,31 @@ func getRemoteSSHKeys(url string) ([]string, error) {
 }
 
 func getFormattedServerURL(addr string) (string, error) {
-	ipErr := checkIP(addr)
-	domainErr := checkDomain(addr)
+	realAddr := addr
+	port := rancherManagementPort
+
```
Comment (bk201) [inline pkg/console/util.go:283]: Just wonder if we can leverage the package: https://pkg.go.dev/net/url

File: pkg/console/util.go [hunk 2]
```diff
@@ -277,12 +278,31 @@ func getRemoteSSHKeys(url string) ([]string, error) {
...
+	if strings.HasPrefix(realAddr, https) {
+		realAddr = strings.TrimPrefix(realAddr, "https://")
+	}
+	if strings.Contains(realAddr, ":") {
+		splitAddr := strings.Split(realAddr, ":")
+		realAddr = splitAddr[0]
+		port = splitAddr[1]
```
Comment (futuretea) [inline pkg/console/util.go:290]: will panic when addr == ""

**Review by Vicente-Cheng — APPROVED** [review]
LGTM, thanks!

File: pkg/console/util_test.go [hunk 3]
```diff
@@ -79,17 +79,11 @@ func TestGetFormattedServerURL(t *testing.T) {
 			output: "https://example.org:443",
 			err:    nil,
 		},
-		{
-			Name:   "invalid ip",
-			input:  "1.2.3.4/",
```
Comment (futuretea) [inline pkg/console/util_test.go:84]: Does RKE2 support this format？

Comment (masteryyh) [reply pkg/console/util_test.go:84]: It's a silly mistake for not checking path in url, added now.

File: pkg/console/util.go [hunk 4]
```diff
@@ -277,12 +278,35 @@ func getRemoteSSHKeys(url string) ([]string, error) {
...
-	return fmt.Sprintf("https://%s:%s", addr, rancherManagementPort), nil
+
+	if parsedUrl.Path != "" {
+		return "", fmt.Errorf("path is not allowed in management address: %s", parsedUrl.Path)
+	}
+
+	port := parsedUrl.Port()
+	if port == "" {
+		parsedUrl.Host += ":443"
+	}
```
Comment (futuretea) [inline pkg/console/util.go:307]:
```go
else if port != "443" {

}
```

Does RKE2 support custom port?
Refer to the logs in https://github.com/harvester/harvester/issues/2737#issue-1362758670
It seems that RKE2 only support 8443 and 443.
And for Harvester, we only use 443.

Comment (masteryyh) [reply pkg/console/util.go:307]: So might be better to add a port verification if not using 443 or 8443?

Comment (futuretea) [reply pkg/console/util.go:307]: just need make sure the port is 443

Comment (futuretea) [reply pkg/console/util.go:307]: 8443 was close by Harvester for security

**Review by guangbochen — APPROVED** [review]
LGTM, thanks.

**Review by futuretea — APPROVED** [review]
LGTM
```yaml
scheme_version: 1
server_url: https://192.168.5.40
token: token             # Replace with a desired token
os:
  hostname: harv-dev-node2
  password: password     # Replace with a your password
  ntp_servers:
  - ntp.aliyun.com
  dns_nameservers:
  - 192.168.5.1
install:
  mode: join
  management_interface:       # The management bond name. This is mandatory.
    interfaces:
    - name: ens33
    default_route: true
    method: dhcp
    bond_options:
      mode: balance-tlb
      miimon: 100
  device: /dev/sda
  iso_url: http://192.168.5.79/_/harvester-pxe/artifacts/master/harvester-amd64.iso
```
image
image

image

**Pull request merged**